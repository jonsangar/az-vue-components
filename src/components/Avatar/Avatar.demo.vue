<template lang="pug">
  .avatarDemo
    Demo(title="Uso Básico")
      az-avatar(src="https://i.pravatar.cc/200")
      az-avatar(name="John Doe")
      az-avatar

    Demo(title="Fallbacks")
      az-avatar.inline-block(src="/image-not-found.jpg" name="John Doe")
      az-avatar.inline-block(src="/image-not-found.jpg")

    Demo(title="Clickable")
      az-avatar(name="John Doe" is-link)
      az-avatar(is-link)

    Demo(title="Tamaños")
      .margin-bottom
        az-avatar(src="https://i.pravatar.cc/200" size="7rem")
        az-avatar(src="https://i.pravatar.cc/200")
        az-avatar(src="https://i.pravatar.cc/200" size="3rem")
      div
        az-avatar(name="John Doe" size="7rem")
        az-avatar(name="John Doe")
        az-avatar(name="John Doe" size="3rem")

    Demo(title="Personalizar colores")
      az-avatar(name="John Doe" background-color="var(--color-brand)" text-color="var(--color-brand-darker)" is-link)
      az-avatar(background-color="var(--color-brand)" text-color="var(--color-brand-darker)" is-link)

    Demo.no-margin-bottom(title="Personalizar icono")
      az-avatar(icon-name="palm" is-link)
</template>

<script>
export default {};
</script>

<style lang="scss">
.avatarDemo {
  min-height: 100vh;

  .az-avatar {
    display: inline-block;
    margin-right: simple();
  }
}
</style>
