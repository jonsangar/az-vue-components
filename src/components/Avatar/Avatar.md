---
device: true
---

# Avatar

### Instalación

```javascript
import { Avatar } from 'didor';

Vue.use(Avatar);
```

---

### Uso

#### Uso básico

``` html
<az-avatar src="path/to/image.jpg"/>
<az-avatar name="John Doe"/>
<az-avatar/>
```

#### Fallbacks

Si el link de la imagen está roto, se muestran las iniciales (si se ha pasado un nombre), o se muestra el icono de fondo.

``` html
<az-avatar src="/image-not-found.jpg" name="John Doe"/>
<az-avatar src="/image-not-found.jpg"/>
```

#### Clickable

Si queremos tener un feedback cuando hacemos click sobre la imagen.

``` html
<az-avatar name="John Doe" is-link/>
<az-avatar is-link/>
```

#### Tamaños

``` html
<az-avatar src="path/to/image.jpg" size="7rem"/>
<az-avatar src="path/to/image.jpg"/>
<az-avatar src="path/to/image.jpg" size="3rem"/>


<az-avatar name="John Doe" size="7rem"/>
<az-avatar name="John Doe"/>
<az-avatar name="John Doe" size="3rem"/>
```

#### Personalizar colores

Si queremos tener un feedback cuando hacemos click sobre la imagen.

``` html
<az-avatar
  name="John Doe"
  background-color="var(--color-brand)"
  text-color="var(--color-brand-darker)"
  is-link/>
<az-avatar
  background-color="var(--color-brand)"
  text-color="var(--color-brand-darker)"
  is-link/>
```

#### Personalizar icono

Si queremos tener un feedback cuando hacemos click sobre la imagen.

``` html
<az-avatar icon-name="palm" is-link/>
```


---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| src | Ruta de la imagen | `String` | -- |
| name | Nombre del usuario | `String` | -- |
| size | Tamaño del avatar | `String` | `5rem` |
| backgroundColor | Color de fondo | `String` | `var(--color-white)` |
| textColor | Color del texto | `String` | `var(--color-gray4)` |
| iconName | Nombre del icono | `String` | `user` |
| isLink | Indica si se puede hacer click sobre el | `Boolean` | `false` |

### Eventos

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| click | Se lanza cuando cambia hacemos click en el avatar | -         |
