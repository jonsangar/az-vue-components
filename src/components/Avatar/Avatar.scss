.az-avatar {
  $rootClass: &;

  @include shadow(3);

  position: relative;
  overflow: hidden;
  font-weight: font-weight(bold);
  text-align: center;
  user-select: none;
  border-radius: 50%;

  &__image {
    position: relative;
    z-index: 3;
    width: 100%;
    height: 100%;
    background-repeat: no-repeat;
    background-size: cover;
  }

  &__initials {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 2;
    width: 100%;
    height: 100%;
    font-size: 0.5em;
  }

  &__icon {
    position: absolute;
    top: 50%;
    left: 50%;
    z-index: 1;
    width: 0.9em;
    height: 0.9em;
    margin: 0 !important;
    transform: translate(-50%, -40%);
  }

  &--link {
    cursor: pointer;

    &:active::after {
      position: absolute;
      top: 0;
      left: 0;
      z-index: 4;
      width: 100%;
      height: 100%;
      content: '';
      background-color: rgba(0, 0, 0, 0.1) !important;
    }
  }
}
