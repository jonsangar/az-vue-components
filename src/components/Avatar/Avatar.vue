<template lang="pug">
  .az-avatar(@click="onClick" :style="avatarStyle" :class="avatarClass")
    az-icon.az-avatar__icon(:name="iconName")
    .az-avatar__initials(v-if="name" :style="avatarInitialStyle") {{ initials }}
    .az-avatar__image(v-if="src" :style="avatarImageStyle")
</template>

<script>
export default {
  props: {
    src: {
      type: String,
      required: false,
    },
    name: {
      type: String,
      required: false,
    },
    size: {
      type: String,
      default: '5rem',
    },
    backgroundColor: {
      type: String,
      default: '#fff',
    },
    textColor: {
      type: String,
      default: 'var(--color-gray4)',
    },
    iconName: {
      type: String,
      default: 'user-fill',
    },
    isLink: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    initials() {
      if (!this.name) return '';
      return this.name
        .split(' ')
        .map(x => x.charAt(0))
        .join('')
        .substr(0, 2)
        .toUpperCase();
    },

    avatarClass() {
      return [this.isLink ? 'az-avatar--link' : ''];
    },

    avatarStyle() {
      return {
        backgroundColor: this.backgroundColor,
        color: this.textColor,
        width: this.size,
        height: this.size,
        lineHeight: this.size,
        fontSize: this.size,
      };
    },

    avatarImageStyle() {
      return {
        backgroundImage: `url(${this.src})`,
      };
    },

    avatarInitialStyle() {
      return {
        backgroundColor: this.backgroundColor,
      };
    },
  },

  methods: {
    onClick(event) {
      this.$emit('clic', event);
    },
  },
};
</script>

<style src="./Avatar.scss" lang="scss"></style>
