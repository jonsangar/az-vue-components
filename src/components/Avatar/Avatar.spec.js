import { mount, createLocalVue } from '@vue/test-utils';
// Importamos librerÃ­as de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Avatar from './Avatar.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Avatar.vue', () => {
  const build = () => {
    const wrapper = mount(Avatar, {
      localVue
    })

    return {
      wrapper,
    }
  }

  it('Renderiza correctamente el compenente avatar con estilos por defecto',async () => {
    const { wrapper } = build();
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot()

    const avatar = wrapper.find('.az-avatar');

    // Contiene un elemento avatar
    expect(avatar.isVisible()).toBe(true);
    // Por defecto contiene los estilo del avatar
    expect(avatar.attributes().style).toContain('background-color: rgb(255, 255, 255); width: 5rem; height: 5rem; line-height: 5rem; font-size: 5rem;');

    const name = wrapper.find('.az-avatar__initials');
    const image = wrapper.find('.az-avatar__image');
    // No contiene la clase de las iniciales del nombre
    expect(name.exists()).toBe(false);
    // No contiene la clase de la imagen
    expect(image.exists()).toBe(false);
    // No es link
    expect(wrapper.find('.az-avatar--link').exists()).toBe(false)
  });
});

describe('Avatar.vue', () => {
  const build = () => {
    const wrapper = mount(Avatar, {
      localVue,
      propsData: {
        src: '@/assets/airzoneWhite.svg',
        name: 'Rafa Montes',
        size: '7rem',
        backgroundColor: '#000',
        textColor: 'var(--color-gray5)',
        isLink: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Renderiza correctamente el compenente avatar con estilos por defecto',async () => {
    const { wrapper } = build();
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot()
    const avatar = wrapper.find('.az-avatar');
    const name = wrapper.find('.az-avatar__initials');
    const image = wrapper.find('.az-avatar__image');
    // Contiene un elemento avatar
    expect(avatar.isVisible()).toBe(true);
    // Los estilos pasados por props se renderizan correctamente
    expect(avatar.attributes().style).toContain('background-color: rgb(0, 0, 0); width: 7rem; height: 7rem; line-height: 7rem; font-size: 7rem;');
    // Contiene la clase y las iniciales del nombre
    expect(name.isVisible()).toBe(true);
    expect(name.text()).toBe('RM');
    // Contiene la clase y los estilos de la imagen
    expect(image.isVisible()).toBe(true);
    expect(image.attributes().style).toContain('background-image: url(@/assets/airzoneWhite.svg);');
    // Es link
    expect(wrapper.find('.az-avatar--link').exists()).toBe(true)
  });

  it('Comprueba que emite el evento clic', async () => {
    const { wrapper } = build()
    await wrapper.vm.$nextTick()
    const e = { currentTarget: true, target: true }
    await wrapper.vm.onClick(e)
    expect(wrapper.emitted().clic).toBeTruthy()
    expect(wrapper.emitted().clic.length).toBe(1)
    expect(wrapper.emitted().clic[0]).toEqual([e])
  });
});
