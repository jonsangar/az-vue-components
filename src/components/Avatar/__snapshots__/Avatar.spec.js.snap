// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar.vue Renderiza correctamente el compenente avatar con estilos por defecto 1`] = `
<div class="az-avatar" style="background-color: rgb(255, 255, 255); width: 5rem; height: 5rem; line-height: 5rem; font-size: 5rem;"><svg version="1.1" viewBox="0 0 16 16" class="az-avatar__icon az-icon az-fill"></svg>
  <!---->
  <!---->
</div>
`;

exports[`Avatar.vue Renderiza correctamente el compenente avatar con estilos por defecto 2`] = `
<div class="az-avatar az-avatar--link" style="background-color: rgb(0, 0, 0); width: 7rem; height: 7rem; line-height: 7rem; font-size: 7rem;"><svg version="1.1" viewBox="0 0 16 16" class="az-avatar__icon az-icon az-fill"></svg>
  <div class="az-avatar__initials" style="background-color: rgb(0, 0, 0);">RM</div>
  <div class="az-avatar__image" style="background-image: url(@/assets/airzoneWhite.svg);"></div>
</div>
`;
